url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/compiler/XTreeCompiler.java

```diff
@@ -0,0 +1,1298 @@
+				}
+			}
+		}
+		if (backSave) {
+			part.set(this, code);
+			if (pop)
+				code.addInstruction(operatorPrefixSuffix, XOpcode.POP_1);
+		} else {
+			part.setupAndGet(this, code);
+		}
+		for (; i >= 0; i--) {
+			XOperator o = operatorPrefixSuffix.prefix.get(i);
+			code.addInstruction(operatorPrefixSuffix, o.opcode);
+		}
+	}
+
+	@Override
+	public void visitIdent(XTreeIdent ident) {
+		treePart = new XCompiledPartVar(ident, scope, ident, XVarAccess.NONE);
+	}
+
+	@Override
+	public void visitIndex(XTreeIndex index) {
+		XCompiledPart array = visit(index.array);
+		XCompiledPart ind = visit(index.index);
+		treePart = new XCompiledPartIndex(index, array, ind);
+	}
+
+	@Override
+	public void visitInstanceof(XTreeInstanceof _instanceof) {
+		XCodeGen code = getCodeExpr(_instanceof);
+		visit(_instanceof.statement).setupAndGet(this, code);
+		visit(_instanceof.type).setupAndGet(this, code);
+		code.addInstruction(_instanceof, XOpcode.INSTANCEOF);
+
+	}
+
+	@Override
+	public void visitIssubclass(XTreeIssubclass issubclass) {
+		XCodeGen code = getCodeExpr(issubclass);
+		visit(issubclass.statement).setupAndGet(this, code);
+		visit(issubclass.type).setupAndGet(this, code);
+		code.addInstruction(issubclass, XOpcode.ISDERIVEDOF);
+	}
+
+	@Override
+	public void visitTypeof(XTreeTypeof typeof) {
+		XCodeGen code = getCodeExpr(typeof);
+		visit(typeof.statement).setupAndGet(this, code);
+		code.addInstruction(typeof, XOpcode.TYPEOF);
+	}
+
+	@Override
+	public void visitSwitch(XTreeSwitch _switch) {
+		XCodeGen code = getCode(_switch);
+		XJumpTarget _break = new XJumpTarget();
+		this.scope = new XScope(this.scope, true, _break, null, labels, 0);
+		labels = null;
+		visit(_switch.statement).setupAndGet(this, code);
+		XInstSwitch s = new XInstSwitch(_switch.position.position.line);
+		code.addInstruction(s);
+		for (XTreeCase _case : _switch.cases) {
+			XJumpTarget jumpTo = new XJumpTarget();
+			code.addInstruction(_switch, jumpTo);
+			if (_case.key instanceof XTreeConstant) {
+				XTreeConstant c = (XTreeConstant) _case.key;
+				if (!s.putIfNonExist(c.value, jumpTo)) {
+					addDiagnostic(_case.key, "case.duplicated.key", c.value);
+				}
+			} else if (_case.key == null) {
+				if (s._default == null) {
+					s._default = jumpTo;
+				} else {
+					addDiagnostic(_case.key, "case.duplicated.default");
+				}
+			} else {
+				addDiagnostic(_case.key, "case.key.no.constant");
+			}
+			visit(_case, _case.block).asStatement(this, code);
+		}
+		if (s._default == null) {
+			s._default = _break;
+		}
+		code.addInstructionB(_switch, XOpcode.POP, this.scope.getLocalsCount());
+		code.addInstruction(_switch, _break);
+		this.scope = this.scope.getParent();
+	}
+
+	@Override
+	public void visitCase(XTreeCase _case) {
+		throw new AssertionError();
+	}
+
+	@Override
+	public void visitConstant(XTreeConstant constant) {
+		treePart = new XCompiledPartConst(constant, constant.value);
+	}
+
+	@Override
+	public void visitCall(XTreeCall call) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 23. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 1146-1166 into a separate method.